.github/workflows/ci.yml

```diff
@@ -73,6 +73,16 @@ jobs:
       runner-env: ${{ needs.setup.outputs.runner-env }}
       write-caches: ${{ needs.setup.outputs.write-caches }}
 
+  build_all_arm64:
+    needs: setup
+    if: contains(fromJson(needs.setup.outputs.enabled-jobs), 'build_all_arm64')
+    uses: ./.github/workflows/build_all.yml
+    with:
+      runner-group: ${{ needs.setup.outputs.runner-group }}
+      runner-env: ${{ needs.setup.outputs.runner-env }}
+      write-caches: ${{ needs.setup.outputs.write-caches }}
+      machine-type: arm64
```

Comment:
> Builds now, but very slow...

We can test how much the caching helps with that. You can flip 'write-caches' to always enabled so presubmit runs populate the cache. By default, only postsubmit runs write to the cache (so presubmit runs on less trusted/stable code don't use storage space and "pollute" the remote cache).

Other things to check would be CPU core count and HDD/SSD setup on the machine.

Reply:
Keeping the discussion on a thread (wish GitHub handled this differently...)

> Veeery slow. Would it be possible to up the machine to "t2a-standard-16" (or even more)?

> Or is it so that ccache need to "fill it's cache" ..?

For builds of the compiler, we should use the largest machine we can find (within reason), _and_ use caching. Building LLVM on a 2 core machine can take hours, for example. We can configure the build a _bit_ to limit how much needs to be compiled (e.g. turning off LLVM targets like RISC-V and Wasm, turning off CUDA and Vulkan), but it's nice to get as much coverage as possible if we're already commiting to a long build.

Reply:
You mean there should be two more runners, one x86 and one arm64, that handles the build_all jobs? Makes sense to me...

OTOH, the regular build job finished in 7m. Let's me try writing the cache.

Reply:
Here's the first run with `write-cache` enabled:

https://github.com/openxla/iree/actions/runs/6584805240/job/17890046166?pr=15237

Then I re-ran:

https://github.com/openxla/iree/actions/runs/6586574344/job/17895211068?pr=15237

Unfortunately no significant change in build time. I must be missing something wrt ccache usage, this should be all in cache right?

We're using the `t2a-standard-8` runner which has 8 vCPU's ([T2A machines](https://cloud.google.com/compute/docs/general-purpose-machines#t2a_machines)). We can potentially up that to 48 vCPU's, but that still don't explain the ~5m compile time on `e2-medium` vs. ~70m on `t2a-standard-8`. I'm not sure how the `e2-medium` burst works too. Anyways, this also makes me believe the issue has to do with ccache, not the machine. The arm64 docker is using and older version of ccache, 3.7. (the Ubuntu 20.04 distro version), but I doubt that would make that big of a difference though... 

I do some digging in the logs and read up on ccache. Any ideas are most welcome :smile:

Reply:
> You mean there should be two more runners, one x86 and one arm64, that handles the build_all jobs? Makes sense to me...

Not sure I follow here. I'm suggesting that we we should use the largest arm64 machine we can reasonably find when building the compiler. I don't know what the options there are - maybe these https://cloud.google.com/compute/docs/general-purpose-machines#t2a_machine_types ? No local SSD on any of those, but maybe "Zonal SSD (performance) Persistent Disk (`pd-ssd`)" is good?

Another option if we don't necessarily need to build and run the compiler on arm64 but still want to run the e2e tests on that platform would be to plug into the `cross_compile_and_test` job. That runs the compiler on a host (x86 Linux) platform and produces artifacts (.vmfb files produced by `iree-compile`) for using in e2e tests on target platforms like riscv and Android (Android is a little special and uses `build_and_test_android` though). Given the specs listed on the cloud page, we should be able to build the full compiler and not treat this as a cross compile job though.

---

RE ccache, if you have a development machine, debugging locally before with the remote cache could be easier. The remote setup looks correct (main thing is using a unique "CCACHE_NAMESPACE"). For local debugging, https://ccache.dev/manual/4.8.3.html#_cache_debugging has general info. I typically look at cache size, hit rate, cacheable calls, and then the debug logs to see if something in the configuration (choice of compiler, flags used, build platform) is unsupported or misconfigured.

Reply:
I chose `t2a-standard-8` to be more cost effective initially. We can change to `t2a-standard-48` if that is needed to achieve a reasonable build time. Our x86 build machine is `n1-standrd-96`, which is quite a powerful machine to compare with, but 70mins on the current `t2a-standard-8` seems to be abnormally long to me.

If you have T2A VM, what is the build time on your VM? I can also test the build on my side to see why it takes so long.

For ccache, IIUC it is expected to be 0 hit because we currently don't upload ccache from presubmit (this PR) to avoid cache poisoning. So every build in this PR will be a new build. The ccache will be updated once it is built on the main branch.

Reply:
To clarify about the ccache, you can enable `write-cache` in presubmit but presubmit runner won't have permissions to write the ccache storage currently.

Reply:
> To clarify about the ccache, write-cache can be enabled in presubmit but presubmit runner won't have permissions to write the ccache storage currently.

Ok, I see. That explains why I got 0% cache hits. I looked for error messages related to writing remote ccache, so I assumed it worked.  

> Not sure I follow here. I'm suggesting that we we should use the largest arm64 machine we can reasonably find when building the compiler.

I thought [these runners](https://github.com/openxla/iree/blob/66d6968bb1b6dda3a98d651a6a758f6879af69f3/build_tools/github_actions/runner/gcp/create_image.sh#L40-L42) were the only runners currently used in this project. My question was whether you were suggesting to add another pair of "compile runners" to do the heavy lifting?

> We can test the build time with ccache on T2A VM locally to know the speedup we will see once this PR is merged. If locally it still takes too long to build, we will need to find other solutions (e.g. have a larger T2A runner).

That would be awesome. I suspect the compile time will be acceptable with ccache properly written.  

> If you have T2A VM, what is the build time on your VM? I can also test the build on my side to see why it takes so long.

> RE ccache, if you have a development machine, debugging locally before with the remote cache could be easier.

I use AWS machines, `c6g.16xlarge` and `c6i.16xlarge`. Without ccache, it takes about 15min. These are 64 vCPU machines, so with a crude translation to the  `t2a-standard-8` runner (8 vCPUs), the 75min compile time is reasonable. With ccache properly written, I expect/hope the compile time will drop to an acceptable level...

Reply:
> With ccache properly written, I expect/hope the compile time will drop to an acceptable level...

I tried this locally and the cache miss rate goes down to only ~45% from 100%. But it was still to slow unfortunately.

When upgrading to ccache 4.7.4, things look a lot better.  So I probably need to build ccache from source [here](https://github.com/openxla/iree/blob/446f96eb896dcc5a1d29ff0af3363d8df1fe7021/build_tools/docker/context/install_ccache.sh#L24-L26) instead, where I used the distro version. I'll make a PR.

Then I need your help to push the new image to the gcp registry. Then use that new docker...

Reply:
Thanks a lot for the testing! Sorry I'm a little bit occupied this week but I'll try to help ASAP

Reply:
I created PR #15307. Let me know when you have time to push the new docker image to GCS. Then I'll update this PR with the new docker image.

Reply:
I created #15308 to fix the issue mentioned [here](https://github.com/openxla/iree/pull/15237#discussion_r1368573696)

Reply:
Hi @freddan80 thanks for waiting. Once you merge #15307 I will push the docker image and post the SHA here.

Reply:
Oh, nvm, I can build image with your PR. Please replace `base-arm64@sha256:31cdb58f556e415c3c65afae31ae9c54e43c1eb3dcd845f3d107abd9b368ff3d` with `base-arm64@sha256:942d01a396f81bff06c5ef4643ba3c9f4500a09010cd30d9ed9be691ddaf1353`

Reply:
You should be able to propose those changes yourself, either by suggesting edits and accepting them or by pushing to the PR branch on the fork @pzread 
![image](https://github.com/openxla/iree/assets/4010439/22c08954-2b28-4b27-a426-3d23e4e2b68c)

(I can also help here, LMK)

Reply:
Thanks for the tips. Done.

Reply:
How about we start these new jobs as postsubmit-only, while we monitor the build times and cache performance?

You can do that by adding the job name(s) to this list: https://github.com/openxla/iree/blob/fcdddcb8a79a30e5d32018ab3a47092f5e44534e/build_tools/github_actions/configure_ci.py#L118-L128

Reply:
LGTM once the job is switched to postsubmit only (for now, while we monitor build stability and compile time)

Reply:
Done